### ISSUE TITLE
`ring_simplify` should not instantiate evars

### ISSUE THREAD

**Issue by samuelgruetter** [coq/coq#14923]
```coq
Require Import ZArith Ring.
Open Scope Z_scope.

Goal forall program_base pos pos0 L,
    exists e, program_base + pos + (pos0 - pos) + L = program_base + e.
Proof.
  intros. eexists.
  ring_simplify.
```

In this example, `ring_simplify` turns

```
  program_base, pos, pos0, L : Z
  ============================
  program_base + pos + (pos0 - pos) + L = program_base + ?e
```

into

```
  program_base, pos, pos0, L : Z
  ============================
  program_base + pos0 + L = program_base + pos0
```

i.e., it instantiates the evar `?e`, which turned a provable goal into an unprovable goal.

Desired behavior: `ring_simplify` should never instantiate evars.

Coq version: 8.13.0

**Comment by SkySkimmer** [comment]
The rewrite at https://github.com/coq/coq/blob/5b4ea258027782d96cb040557b20f9b3173940e0/theories/setoid_ring/Ring_tac.v#L423 gets `H : program_base + ?e = program_base + ?e`.
Indeed after `eexists` doing `match goal with |- _ = ?rhs => rewrite (eq_refl : rhs = rhs) end.` causes a similar unification (not the same because the `program_base + _` subterm it encounters isn't the same).

PS: I don't understand why but 
~~~coq
let e := open_constr:(_) in 
eexists e;
ring_simplify.
~~~
does not simplify the rhs.

**Comment by JasonGross** [comment]
Perhaps `ring_simplify` should pose all evars first a la https://github.com/mit-plv/fiat-crypto/blob/master/src/Util/Tactics/SetEvars.v and then subst them a la https://github.com/mit-plv/fiat-crypto/blob/master/src/Util/Tactics/SubstEvars.v ?

**Comment by JasonGross** [comment]
> PS: I don't understand why but
> 
> ```coq
> let e := open_constr:(_) in 
> eexists e;
> ring_simplify.
> ```
> 
> does not simplify the rhs.

Presumably because this creates the evar before the intros (now done implicitly by `eexists`) and hence it can no longer be instantiated by `pos + (pos0 - pos) + L`

**Comment by SkySkimmer** [comment]
>Presumably because this creates the evar before the intros (now done implicitly by eexists) and hence it can no longer be instantiated by pos0 + L

No, I left the intros in (ie the script I tried was `intros. let e := ...`)

**Comment by JasonGross** [comment]
I've minimized this difference in behavior to https://github.com/coq/coq/issues/15520.  Unfortunately fixing that bug to have sane behavior will result in this bug occurring regardless of whether or not a `;` is used.
